linera-core/src/client/mod.rs

```diff
@@ -832,31 +832,47 @@ impl<Env: Environment> Client<Env> {
                     warn!("Skipping received certificate from past epoch {epoch:?}");
                 }
                 CheckCertificateResult::New => {
-                    downloaded_heights
+                    certificates
                         .entry(sender_chain_id)
-                        .and_modify(|h| *h = height.max(*h))
-                        .or_insert(height);
-                    certificates.push(confirmed_block_certificate);
+                        .or_insert_with(BTreeMap::new)
+                        .insert(height, confirmed_block_certificate.clone());
                 }
             }
         }
 
         // Increase the tracker up to the first position we haven't downloaded.
         for entry in remote_log {
-            if downloaded_heights
+            if certificates
                 .get(&entry.chain_id)
-                .is_some_and(|h| *h >= entry.height)
+                .is_some_and(|certs| certs.contains_key(&entry.height))
             {
                 tracker += 1;
             } else {
                 break;
             }
         }
 
+        for (sender_chain_id, certs) in &mut certificates {
+            if !certs
+                .values()
+                .last()
+                .is_some_and(|cert| cert.block().recipients().contains(&chain_id))
+            {
+                warn!(
+                    "Skipping received certificates from chain {sender_chain_id:.8}:
+                    No messages for {chain_id:.8}."
+                );
+                certs.clear();
+            }
+        }
+
         Ok(ReceivedCertificatesFromValidator {
             public_key: remote_node.public_key,
             tracker,
-            certificates,
+            certificates: certificates
+                .into_values()
+                .flat_map(BTreeMap::into_values)
+                .collect(),
```

Comment:
So, for a given chain, we can take certificates of different heights, not just the last one.

Reply:
The last one should be the one that sent the latest message to our chain.

_Currently_ we download all its descendants, but soon (#3969) we will download only the ones that sent messages.